File: Renderers/TestRenderer.swift
typealias TestImage = [String]

/// 'Renders' images into an array of strings that are easy to test
class TestRenderer: Renderer, ImageRenderer {
  typealias ImageType = TestImage
  var currentImage: ImageType = ImageType()

  func line(_ a: Point, _ b: Point) {
    currentImage.append("line: \(a), \(b)")
  }

  func arc(_ a: Point, _ b: Point) {
    currentImage.append("arc: \(a), \(b)")
  }

  func circle(_ center: Point, radius: Double) {
    currentImage.append("circle: \(center), \(radius)")
  }

  func bezier(_ a: Point, _ cp1: Point, _ cp2: Point, _ b: Point) {
    currentImage.append("bezier: \(a), [\(cp1), \(cp2)], \(b)")
  }

  func image(_ image: ImageType) {
    self.currentImage.append(contentsOf: image)
  }

  func moveTo(_ point: Point) {
    currentImage.append("move: \(point)")
  }

  func stroke(_ lineWidth: Double) {
    currentImage.append("stroke")
  }

  func fill() {
    currentImage.append("fill")
  }

  func shadowOn() {
    currentImage.append("shadow on")
  }

  func shadowOff() {
    currentImage.append("shadow off")
  }

  func clear() {
    self.currentImage = []
  }

  func color(_ color: Color) {
    currentImage.append("color")
  }

  func placeImage(start: Point, width: Double, height: Double, name: String) {

  }
}

File: clients/browser/app/renderers/TestRenderer.js
export default class TestRenderer {
  constructor() {
    this.image = [];
  }

  currentImage() {
    return _.clone(this.image);
  }

  line(a, b) {
    this.image.push(`line: ${a}, ${b}`);
  }

  arc(a, b) {
    this.image.push(`arc: ${a}, ${b}`);
  }

  circle(center, radius) {
    this.image.push(`circle: ${center}, ${radius}`);
  }

  bezier(a, cp1, cp2, b) {
    this.image.push(`bezier: ${a}, [${cp1}, ${cp2}], ${b}`);
  }

  image(image) {
    this.image.append(image);
  }

  moveTo(point) {
    this.image.push(`move: ${point})`);
  }

  stroke(lineWidth) {
    this.image.push(`stroke (${lineWidth})`);
  }

  fill() {
    this.image.push('fill');
  }

  shadowOn() {
    this.image.push('shadow on');
  }

  shadowOff() {
    this.image.push('shadow off');
  }

  clear() {
    this.image = [];
  }

  color(color) {
    this.image.push(`color ${color}`);
  }

  placeImage(start, width, height, name) {

  }
}
